File: shared/protocol.ts
// wink 协议核心（TS 单一权威实现）
//
// 跨语言确定性铁律：
// - 本文件是协议字节布局的唯一事实源（Rust 端复制 + golden 断言锁死）
// - 所有整数运算必须与 Rust u32 wrapping 语义一致
// - 协议路径禁止 Math.log/Math.random（用 dlog / splitmix32）

// ─────────────────────────────────────────────
// 常量
// ─────────────────────────────────────────────

export const FRAME_MAGIC = 0x57; // 'W'
export const FRAME_VERSION = 0x01;
export const HEADER_LEN = 20; // magic(1)+version(1)+sessionId(2)+seq(4)+k(2)+blockLen(2)+totalLen(4)+payloadFnv(4)

export const FILE_MAGIC = new Uint8Array([0x57, 0x4e, 0x4b, 0x31]); // "WNK1"
export const TEXT_MAGIC = new Uint8Array([0x57, 0x4e, 0x4b, 0x54]); // "WNKT"
export const MANIFEST_MAGIC = new Uint8Array([0x57, 0x4e, 0x4b, 0x4d]); // "WNKM"

export const FILE_HEADER_LEN = 49;
export const MAX_FILE_BYTES = 64 * 1024 * 1024;
export const MAX_FILE_LABEL = `${MAX_FILE_BYTES / 1024 / 1024} MB`;
export const MAX_SOURCE_BLOCKS = 0xffff;
export const MAX_SNIPPET_BYTES = 4 * 1024 * 1024;

// ─────────────────────────────────────────────
// FNV-1a（纯 u32 整数运算，与 Rust wrapping 一致）
// ─────────────────────────────────────────────

export function fnv1a(bytes: Uint8Array): number {
  let h = 0x811c9dc5;
  for (let i = 0; i < bytes.length; i++) {
    h ^= bytes[i]!;
    h = Math.imul(h, 0x01000193);
  }
  return h >>> 0;
}

// ─────────────────────────────────────────────
// splitmix32（确定性 PRNG，纯整数运算）
// JS: Math.imul ≡ Rust: wrapping_mul
// JS: >>> ≡ Rust: >> (u32)
// ─────────────────────────────────────────────

export function splitmix32(seed: number): () => number {
  let s = seed | 0;
  return () => {
    s = (s + 0x9e3779b9) | 0;
    let t = s ^ (s >>> 16);
    t = Math.imul(t, 0x21f0aaad);
    t ^= t >>> 15;
    t = Math.imul(t, 0x735a2d97);
    t ^= t >>> 15;
    return t >>> 0;
  };
}

// ─────────────────────────────────────────────
// 帧头
// ─────────────────────────────────────────────

export interface FrameHeader {
  sessionId: number; // u16
  seq: number; // u32
  k: number; // u16
  blockLen: number; // u16
  totalLen: number; // u32
  payloadFnv: number; // u32
}

export function packFrame(h: FrameHeader, block: Uint8Array): Uint8Array {
  const out = new Uint8Array(HEADER_LEN + block.length);
  const dv = new DataView(out.buffer);
  dv.setUint8(0, FRAME_MAGIC);
  dv.setUint8(1, FRAME_VERSION);
  dv.setUint16(2, h.sessionId, true);
  dv.setUint32(4, h.seq, true);
  dv.setUint16(8, h.k, true);
  dv.setUint16(10, h.blockLen, true);
  dv.setUint32(12, h.totalLen, true);
  dv.setUint32(16, h.payloadFnv, true);
  out.set(block, HEADER_LEN);
  return out;
}

export function parseFrame(
  bytes: Uint8Array,
): { header: FrameHeader; block: Uint8Array } | null {
  if (bytes.length <= HEADER_LEN) return null;
  if (bytes[0] !== FRAME_MAGIC || bytes[1] !== FRAME_VERSION) return null;
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const header: FrameHeader = {
    sessionId: dv.getUint16(2, true),
    seq: dv.getUint32(4, true),
    k: dv.getUint16(8, true),
    blockLen: dv.getUint16(10, true),
    totalLen: dv.getUint32(12, true),
    payloadFnv: dv.getUint32(16, true),
  };
  if (header.k === 0 || header.blockLen === 0 || header.totalLen === 0) return null;
  if (bytes.length !== HEADER_LEN + header.blockLen) return null;
  return { header, block: bytes.subarray(HEADER_LEN) };
}

/** 流标识：任何字段变化即重置（防 sessionId 16 位碰撞静默错乱） */
export function streamIdentity(h: FrameHeader): string {
  return `${h.sessionId}:${h.k}:${h.blockLen}:${h.totalLen}:${h.payloadFnv}`;
}

// ─────────────────────────────────────────────
// 安全文件名
// ─────────────────────────────────────────────

export function safeFileName(name: string): string {
  const base = name.split(/[\\/]/).pop() ?? '';
  const cleaned = base.replace(/[\u0000-\u001f\u007f]/g, '').trim();
  return cleaned === '' || cleaned === '.' || cleaned === '..'
    ? 'transfer.bin'
    : cleaned;
}

File: sender-cli/src/protocol.rs
// wink 协议核心（Rust 端）
//
// 必须与 shared/protocol.ts 逐位一致（golden 向量断言锁定）。
// - u32 运算用 wrapping_*（与 JS Math.imul/>>> 语义一致）
// - f64 运算保持 IEEE-754 默认语义

#![allow(
    clippy::cast_precision_loss,
    clippy::cast_possible_truncation,
    clippy::cast_sign_loss,
    clippy::cast_possible_wrap,
    clippy::many_single_char_names,
    clippy::cast_lossless,
    clippy::items_after_statements
)]

pub const FRAME_MAGIC: u8 = 0x57; // 'W'
pub const FRAME_VERSION: u8 = 0x01;
pub const HEADER_LEN: usize = 20;

pub const FILE_MAGIC: [u8; 4] = [0x57, 0x4e, 0x4b, 0x31]; // "WNK1"
pub const TEXT_MAGIC: [u8; 4] = [0x57, 0x4e, 0x4b, 0x54]; // "WNKT"
pub const MANIFEST_MAGIC: [u8; 4] = [0x57, 0x4e, 0x4b, 0x4d]; // "WNKM"

pub const FILE_HEADER_LEN: usize = 49;
pub const MAX_FILE_BYTES: u64 = 64 * 1024 * 1024;
pub const MAX_SOURCE_BLOCKS: u16 = 0xffff;
pub const MAX_SNIPPET_BYTES: u32 = 4 * 1024 * 1024;

/// FNV-1a（纯 u32 wrapping 运算）
#[must_use]
pub fn fnv1a(bytes: &[u8]) -> u32 {
    let mut h: u32 = 0x811c_9dc5;
    for &b in bytes {
        h ^= b as u32;
        h = h.wrapping_mul(0x0100_0193);
    }
    h
}

/// splitmix32 确定性 PRNG（与 TS 逐位一致）
pub struct SplitMix32 {
    s: u32,
}

impl SplitMix32 {
    #[must_use]
    pub fn new(seed: u32) -> Self {
        Self { s: seed }
    }
    #[must_use]
    pub fn next_u32(&mut self) -> u32 {
        self.s = self.s.wrapping_add(0x9e37_79b9);
        let mut t = self.s ^ (self.s >> 16);
        t = t.wrapping_mul(0x21f0_aaad);
        t ^= t >> 15;
        t = t.wrapping_mul(0x735a_2d97);
        t ^= t >> 15;
        t
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct FrameHeader {
    pub session_id: u16,
    pub seq: u32,
    pub k: u16,
    pub block_len: u16,
    pub total_len: u32,
    pub payload_fnv: u32,
}

#[must_use]
pub fn pack_frame(h: &FrameHeader, block: &[u8]) -> Vec<u8> {
    let mut out = vec![0u8; HEADER_LEN + block.len()];
    out[0] = FRAME_MAGIC;
    out[1] = FRAME_VERSION;
    out[2..4].copy_from_slice(&h.session_id.to_le_bytes());
    out[4..8].copy_from_slice(&h.seq.to_le_bytes());
    out[8..10].copy_from_slice(&h.k.to_le_bytes());
    out[10..12].copy_from_slice(&h.block_len.to_le_bytes());
    out[12..16].copy_from_slice(&h.total_len.to_le_bytes());
    out[16..20].copy_from_slice(&h.payload_fnv.to_le_bytes());
    out[HEADER_LEN..].copy_from_slice(block);
    out
}

#[must_use]
pub fn parse_frame(bytes: &[u8]) -> Option<(FrameHeader, &[u8])> {
    if bytes.len() <= HEADER_LEN {
        return None;
    }
    if bytes[0] != FRAME_MAGIC || bytes[1] != FRAME_VERSION {
        return None;
    }
    let header = FrameHeader {
        session_id: u16::from_le_bytes([bytes[2], bytes[3]]),
        seq: u32::from_le_bytes([bytes[4], bytes[5], bytes[6], bytes[7]]),
        k: u16::from_le_bytes([bytes[8], bytes[9]]),
        block_len: u16::from_le_bytes([bytes[10], bytes[11]]),
        total_len: u32::from_le_bytes([bytes[12], bytes[13], bytes[14], bytes[15]]),
        payload_fnv: u32::from_le_bytes([bytes[16], bytes[17], bytes[18], bytes[19]]),
    };
    if header.k == 0 || header.block_len == 0 || header.total_len == 0 {
        return None;
    }
    if bytes.len() != HEADER_LEN + header.block_len as usize {
        return None;
    }
    Some((header, &bytes[HEADER_LEN..]))
}

#[must_use]
pub fn stream_identity(h: &FrameHeader) -> String {
    format!(
        "{}:{}:{}:{}:{}",
        h.session_id, h.k, h.block_len, h.total_len, h.payload_fnv
    )
}

/// 安全文件名：剥离路径/控制字符（与 TS safeFileName 一致）
#[must_use]
pub fn safe_file_name(name: &str) -> String {
    let base = name.rsplit(['/', '\\']).next().unwrap_or("");
    let cleaned: String = base
        .chars()
        .filter(|c| !c.is_control())
        .collect::<String>()
        .trim()
        .to_string();
    if cleaned.is_empty() || cleaned == "." || cleaned == ".." {
        "transfer.bin".to_string()
    } else {
        cleaned
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fnv1a_empty_is_offset_basis() {
        assert_eq!(fnv1a(&[]), 0x811c_9dc5);
    }

    #[test]
    fn fnv1a_abc_matches_golden() {
        // golden: 0x1a47e90b (TS 实测)
        assert_eq!(fnv1a(b"abc"), 0x1a47e90b);
    }

    #[test]
    fn splitmix32_fixed_sequence() {
        let mut rnd = SplitMix32::new(1234);
        let a: Vec<u32> = (0..10).map(|_| rnd.next_u32()).collect();
        let mut rnd2 = SplitMix32::new(1234);
        let b: Vec<u32> = (0..10).map(|_| rnd2.next_u32()).collect();
        assert_eq!(a, b);
    }

    #[test]
    fn pack_parse_roundtrip() {
        let h = FrameHeader {
            session_id: 1,
            seq: 42,
            k: 8,
            block_len: 64,
            total_len: 512,
            payload_fnv: 0xdeadbeef,
        };
        let block: Vec<u8> = (0..64).collect();
        let frame = pack_frame(&h, &block);
        assert_eq!(frame.len(), HEADER_LEN + 64);
        let (parsed, parsed_block) = parse_frame(&frame).unwrap();
        assert_eq!(parsed, h);
        assert_eq!(parsed_block, block);
    }

    #[test]
    fn parse_rejects_bad() {
        assert!(parse_frame(&[0u8; 5]).is_none());
        let mut bad = vec![0u8; HEADER_LEN + 64];
        bad[0] = 0;
        assert!(parse_frame(&bad).is_none());
    }

    #[test]
    fn safe_name_strips() {
        assert_eq!(safe_file_name("../evil.txt"), "evil.txt");
        assert_eq!(safe_file_name("/abs/path/file"), "file");
        assert_eq!(safe_file_name(".."), "transfer.bin");
        assert_eq!(safe_file_name("正常文件.txt"), "正常文件.txt");
    }
}
